rcli/typing.py

```diff
@@ -0,0 +1,673 @@
+        keyfunc_name = cls._get_fullname(keyfunc)
+        identity = cls._identity
+        try:
+            class BaseClass(type_):
+                """Use a subclass of type_ if it is subclassable."""
+                pass
+        except TypeError:
+            BaseClass = object
+
+        @six.add_metaclass(_ClsReprMeta)
+        class _BoundedSubclass(BaseClass):  # type: ignore
+            """A subclass of a type bounded by a slice."""
+
+            def __new__(cls, __value, *args, **kwargs):
+                # type: (Any, *Any, **Any) -> _T
+                """Return __value cast to _T.
+
+                Any additional arguments are passed as-is to the constructor.
+
+                Args:
+                    __value: A value that can be converted to type _T.
+                    args: Any additional positional arguments passed to the
+                        constructor.
+                    kwargs: Any additional keyword arguments passed to the
+                        constructor.
+                """
+                try:
+                    ret = type_(__value, *args, **kwargs)
+                except TypeError:
+                    ret = __value
+                expected = keyfunc(ret)
+                if bound.start and expected < bound.start:
+                    if keyfunc is not identity:
+                        raise ValueError(
+                            'The value of {}({}) [{}] is below the minimum '
+                            'allowed value of {}.'.format(
+                                keyfunc_name, repr(__value), repr(expected),
+                                bound.start))
+                    raise ValueError(
+                        'The value {} is below the minimum allowed value '
+                        'of {}.'.format(repr(__value), bound.start))
+                if bound.stop and expected > bound.stop:
+                    if keyfunc is not identity:
+                        raise ValueError(
+                            'The value of {}({}) [{}] is above the maximum'
+                            ' allowed value of {}.'.format(
+                                keyfunc_name, repr(__value), repr(expected),
+                                bound.stop))
+                    raise ValueError(
+                        'The value {} is above the maximum allowed value '
+                        'of {}.'.format(repr(__value), bound.stop))
+                return ret
+
+        if keyfunc is not cls._default:
+            _BoundedSubclass.__class_repr__ = '{}.{}[{}, {}, {}]'.format(
+                cls.__module__, cls.__name__, cls._get_fullname(type_),
+                cls._get_bound_repr(bound), keyfunc_name)
+        else:
+            _BoundedSubclass.__class_repr__ = '{}.{}[{}, {}]'.format(
+                cls.__module__, cls.__name__, cls._get_fullname(type_),
+                cls._get_bound_repr(bound))
+        return _BoundedSubclass
+
+    def _get_args(cls, args):
```

Comment:
I feel like "_get_args" isn't the best name for this.

One result of this function call is "type_" I wonder if the verb "`declare`" might be appropriate in its name?

I think "`bounds`" might be a slightly clearer name than "`bound`" for the second returned value.